design/0001-frr.md

```diff
@@ -0,0 +1,402 @@
+  a whole (To be confirmed)
+* This enhancement assumes that FRR will be deployed as a DaemonSet on the
+  Kubernetes cluster. It is also assumed that there is a stable/versioned
+  interface between this DaemonSet and MetalLB. Therefore, if any non-backwards
+  compatible changes are made in this interface, this will need to be managed in
+  the MetalLB ‘speaker’ component.
+* Upgrade/downgrade will remove a node from the MetalLB memberlist by unlabeling
+  the node. At this point the ‘speaker’ component on the node can be stopped and
+  the FRR component can be updated to the new version. It should be noted that
+  as FRR only peers with BGP agents outside the cluster, there is no requirement
+  that each node within the cluster maintains the same version or a compatible
+  version of FRR. However, it is required that each version of FRR is compatible
+  with the BGP agent with which it is peering. After FRR is updated, the
+  ‘speaker’ component can be restarted and the node can be labelled to add it to
+  the memberlist again.
+
+### Version Skew Strategy
+
+Version skew between versions of FRR should not be a concern to Kubernetes or
+MetalLB as long as FRR presents a stable interface to the MetalLB ‘speaker’
+component. This is because FRR will only peer with BGP agents outside of the
+cluster and not with each other.
+
+## Drawbacks
+
+This would couple MetalLB with the FRR community so that it would be somewhat
+dependent on this community for bug-fixes and further feature improvements.
+However, there would be the native BGP agent as a fallback for cluster
+administrators that did not want to deploy FRR.
+
+## Alternatives
+
+* In terms of enhancing the MetalLB native BGP implementation, an alternative
+  would be to invest more effort in adding additional features.
+* In terms of enabling another BGP agent, there may be other approaches that
+  could be investigated but from the current investigation, this seems the least
+  intrusive.
+
+### BGP Stack Alternatives
+
+A number of alternative open-source routing stacks were considered (FRR, BIRD,
+GoBGP) as a first target for integration. They were evaluated across a number of
+categories. GoBGP was discounted due to its relatively limited feature set. For
+example, it does not support [BFD](https://github.com/osrg/gobgp/issues/2006).
+FRR and BIRD are well-known and mature stacks which have been deployed in
+production and have active development communities. FRR was selected as the
+first target for integration for the following reasons:
+
+* Integration options: Both BIRD and FRR can be configured by MetalLB using the
+  CLI and/or configuration files. However, FRR also has the option to be
+  configured with a gRPC interface which can generate Go bindings.
+* Extensibility options: FRR provides additional extensibility via its `dplane`
+  and `FPM` interfaces. This may ease integration with Kubernetes networking
+  providers. There does not appear to be equivalents in BIRD.
```

Comment:
It could be worth elaborating on this a bit to give an example of what type of integration you have in mind here?

Reply:
Sure, done.